drivers: wifi: infineon:  Add SPI support to AIROC WiFi driver

ThreeEights opened the pull request:
Added support for SPI to the Infineon AIROC WiFi driver (drivers/wifi/infineon).  Includes support for 3-wire SPI required by the Raspberry Pi Pico W.

Configuration for the Pico W is updated to activate the WiFi chip on that board.

parthitce commented on dts/bindings/wifi/infineon,airoc-wifi-spi.yaml:
```diff
@@ -0,0 +1,37 @@
+include: [spi-device.yaml, "infineon,airoc-wifi-common.yaml"]
+
+properties:
+  wifi-host-wake-gpios:
```
duplicated from common.

ThreeEights replied:
True.  Turns out I can just add the "required" attribute in infineon,airoc-wifi-spi.yaml.

parthitce commented on dts/bindings/wifi/infineon,airoc-wifi-common.yaml:
```diff
@@ -0,0 +1,20 @@
+      by the WiFi section of AIROC Wi-Fi device.
+    type: phandle-array
+
+  wifi-host-wake-gpios:
```
can be renamed to `interrupt` IMO

ThreeEights replied:
"WIFI_HOST_WAKE" comes from the chip datasheet, and that's how Infineon defined the property.  I didn't want to change it.

parthitce replied:
yeah, same is called `interrupt` in Linux Kernel context, so was my suggestion. prefix `wifi-` is implicit already, so can be removed

ThreeEights replied:
The names:
- wifi_host_wake
- wifi_reg_on
- bt_reg_on
are already in use in the Infineon WiFi and Bluetooth drivers, and match names used throughout modules/hal/infineon and zephyr/modules/hal_infineon, as well as following the naming convention used in their hardware datasheets.  Since I'm not part of Infineon, and I'm only adding a new capability on top of their existing driver, I think it's not my role to modify their names.

parthitce commented on dts/bindings/wifi/infineon,airoc-wifi-common.yaml:
```diff
@@ -0,0 +1,20 @@
+      to the host indicating that the device requires attention.
+    type: phandle-array
+
+  wifi-dev-wake-gpios:
```
can be named as `wakeup-gpios`  IMO

ThreeEights replied:
Again, the name comes from the chip's datasheet.

parthitce reviewed (changes requested):
split the changes. bindings, driver and consumer

soburi commented on modules/hal_infineon/wifi-host-driver/CMakeLists.txt:
```diff
@@ -49,7 +49,11 @@ zephyr_library_sources(${hal_wifi_dir}/WiFi_Host_Driver/src/whd_wifi_p2p.c)
 zephyr_library_sources(${hal_wifi_dir}/WiFi_Host_Driver/src/bus_protocols/whd_bus.c)
 zephyr_library_sources(${hal_wifi_dir}/WiFi_Host_Driver/src/bus_protocols/whd_bus_common.c)
-zephyr_library_sources(${hal_wifi_dir}/WiFi_Host_Driver/src/bus_protocols/whd_bus_sdio_protocol.c)
+if(CONFIG_AIROC_WIFI_BUS_SPI)
```
It is better to use `zephyr_library_sources_ifdef`.

ThreeEights replied:
Done.

soburi commented on drivers/wifi/infineon/airoc_wifi.h:
```diff
@@ -6,16 +6,48 @@
+#if defined(CONFIG_AIROC_WIFI_BUS_SPI_DATA_IRQ_MUX) && defined(CONFIG_BOARD_RPI_PICO)
+#define PICO_BUS_DEVICE DT_PARENT(DT_PARENT(DT_DRV_INST(0)))
+
+#if PICO_BUS_DEVICE == DT_NODELABEL(pio0)
```
This is very strange and violates the independence of the modules.
Can't you put it on the pio-spi side?

ThreeEights commented:
@parthitce - You wrote "split the changes. bindings, driver and consumer".  Do you mean create 3 separate pull requests?  Or do you mean something else?

parthitce commented:
> @parthitce - You wrote "split the changes. bindings, driver and consumer". Do you mean create 3 separate pull requests? Or do you mean something else?

No, split into multiple commits in this PR

ThreeEights committed:
dts: bindings: wifi: Support SPI for Infineon AIROC driver

Additional bindings to configure SPI support.

Added new DTS option for data/IRQ sharing
Clarify default in driver DTS binding

Signed-off-by: Steve Boylan <stephen.boylan@beechwoods.com>

ThreeEights committed:
drivers: wifi: infineon: Add SPI support to AIROC driver

Added support for SPI in 4-wire and 3-wire configurations to
the Infineon AIROC WiFi driver (drivers/wifi/infineon).

Review changes:
Move DT_DRV_COMPAT to common header file
Correct board-specific preprocessor lines
Removed AIROC_MAP_COUNTRY_CODE
Move Pico W configuration details to devicetree
Use pinctrl to manage shared data/interrupt GPIO
Clean up bus selection in Kconfig.airoc
Make SDIO and SPI bus struct independent
Replace LOG_DBG with LOG_ERR
Remove functionally duplicate operation
Remove spurious Kconfig option
Minor cleanup in CMakeLists.txt

Signed-off-by: Steve Boylan <stephen.boylan@beechwoods.com>

ThreeEights committed:
boards: rpi_pico:  WiFi configuration for Pico W

Configure the Raspberry Pi Pico W for WiFi.
Move Pico W configuration details to devicetree
Add pinctrl configurations for data/interrupt sharing
Make memory config selectable
Align devicetree with Linux ordering

Signed-off-by: Steve Boylan <stephen.boylan@beechwoods.com>

ThreeEights commented:
@parthitce - I divided this into 3 commits as you requested.  Please let me know if these meet your expectations.

Unfortunately, these changes depend on a pull request for modules/hal/infineon, which is still being reviewed by Infineon.  I'll keep following that change so this can proceed.

soburi commented on drivers/wifi/infineon/Kconfig.airoc:
```diff
@@ -17,6 +16,24 @@ menuconfig WIFI_AIROC
 
 if WIFI_AIROC
 
+choice
```
Can this configuration replace with `DT_ON_BUS`?

npal-cy replied:
yes looks like it a good idea to select spi/sdio(sdhc) by DT_ON_BUS, so it will simplify bus configuration. For now we need disable one, enable another...:
```
CONFIG_AIROC_WIFI_BUS_SDIO=n
CONFIG_AIROC_WIFI_BUS_SPI=y
```

ThreeEights replied:
@soburi - Looks like most of the cases can be replaced with DT_ON_BUS().  The two I don't know how to address are in CMakeLists.txt, where the appropriate bus interface source file is selected, and making the option AIROC_WIFI_BUS_SPI_DATA_IRQ_MUX available in Kconfig.airoc.  The latter enables logic to change use of the GPIO line in the special case where the SPI data and IRQ are shared, which is the case for the Pico W.
Do you have any suggestions for how to handle those?

soburi replied:
Can it be configured by pinctrl?
IMO, The pin function should not be configured in the wifi driver.

ThreeEights replied:
> yes looks like it a good idea to select spi/sdio(sdhc) by DT_ON_BUS, so it will simplify bus configuration. For now we need disable one, enable another...:
> 
> ```
> CONFIG_AIROC_WIFI_BUS_SDIO=n
> CONFIG_AIROC_WIFI_BUS_SPI=y
> ```

@npal-cy - Yes, that's why I made it a choice.  Now that I know about DT_ON_BUS, yes, that makes things cleaner ... if we can figure out how to configure cmake and the data/IRQ multiplexing.

ThreeEights replied:
> Can it be configured by pinctrl?
> IMO, The pin function should not be configured in the wifi driver.

I struggled with this issue.  On the Pico W, GPIO 24 is used for bidirectional data as well as HOST_WAKE.  In order to make this work, the pin has to be assigned to SIO (function 5) when serving as an interrupt, but switched to the PIO (function 6 or 7, depending) when used for SPI data.  The Pico SPI driver doesn't know about this silliness;  it seems to belong in the layer within the WiFi driver that deals with the SPI bus.  This was the best compromise I could come up with.  I'd be happy to hear any cleaner solutions.

npal-cy replied:
> @npal-cy - Yes, that's why I made it a choice. Now that I know about DT_ON_BUS, yes, that makes things cleaner ... if we can figure out how to configure cmake and the data/IRQ multiplexing.

I means to keep AIROC_WIFI_BUS_SPI/AIROC_WIFI_BUS_SDIO, make them hided and auto enabled by on_bus, something like snipped below (not tested...)
```
config AIROC_WIFI_BUS_SDIO
    bool
    default $(dt_compat_on_bus,$(DT_COMPAT_INFINEON_AIROC_WIFI),sd)
    select SDIO_STACK
	select SDHC

config AIROC_WIFI_BUS_SPI
    bool
    default $(dt_compat_on_bus,$(DT_COMPAT_INFINEON_AIROC_WIFI),spi)
    select SPI

```

soburi replied:
@ThreeEights 

> Do you have any suggestions for how to handle those?

We can use `pinctrl_apply_state` with a private state.

https://github.com/zephyrproject-rtos/zephyr/blob/2f47de60a0a7038eed1bfe11ced16ae51444434e/drivers/sdhc/imx_usdhc.c#L176

ThreeEights replied:
@soburi - I think I've come up with a better solution.  I've changed most of what had been Kconfig settings to rely instead on devicetree.  See what you think of this version - changes in:
```
dts/bindings/wifi/infineon,airoc-wifi-spi.yaml
boards/raspberrypi/rpi_pico/rpi_pico_rp2040_w_defconfig
boards/raspberrypi/rpi_pico/rpi_pico_rp2040_w.dts
drivers/wifi/infineon/Kconfig.airoc
drivers/wifi/infineon/airoc_whd_hal_spi.c
drivers/wifi/infineon/airoc_wifi.h
```

soburi replied:
@ThreeEights 

The overall bus structure has been improved by organizing it.
However, this pinctrl topic is separate from this and is the last piece of the puzzle.
It should be applied to avoid leaving the rp2040-dependent code here.

ThreeEights replied:
@soburi @npal-cy - I reworked control of the GPIO shared between data and the HOST_WAKE interrupt to use pinctrl instead of direct control.  Since that is specific to the SPI code in the Infineon driver, the pinctrl configurations must be handled at that level, which makes the RPI Pico dts look a bit peculiar.  Those changes affect:
```
boards/raspberrypi/rpi_pico/rpi_pico_rp2040_w.dts
drivers/wifi/infineon/airoc_whd_hal_spi.c
drivers/wifi/infineon/airoc_wifi.c
drivers/wifi/infineon/airoc_wifi.h
```
I also modified drivers/wifi/infineon/Kconfig.airoc so the bus selection now depends only on the dt_compat_on_bus extension.

ThreeEights commented on drivers/wifi/infineon/airoc_whd_hal_spi.c:
```diff
@@ -0,0 +1,296 @@
+#include <hardware/gpio.h>
+#include <zephyr/drivers/spi.h>
+
+#define DT_DRV_COMPAT infineon_airoc_wifi
```
Good point.  I've moved it as you suggested.

ThreeEights commented on drivers/wifi/infineon/airoc_whd_hal_spi.c:
```diff
@@ -0,0 +1,296 @@
+#include <bus_protocols/whd_bus_spi_protocol.h>
+#include <bus_protocols/whd_spi.h>
+#include <whd_wifi_api.h>
+#include <hardware/gpio.h>
```
Ooops!  Yes, that's Pico W specific.  I wrapped it in #if/#endif.

ThreeEights commented on drivers/wifi/infineon/airoc_whd_hal_common.h:
```diff
@@ -0,0 +1,60 @@
+/** Map Zephyr country codes (2 octets) to whd_country_code_t. */
+#if defined(CONFIG_AIROC_WIFI_COUNTRY)
+#if defined(CONFIG_BIG_ENDIAN)
+#define AIROC_MAP_COUNTRY_CODE(code) ((code[1]) + (code[0] << 8))
```
Sorry, that's a vestige of an issue that was fixed a long time ago.  I took it out.

ThreeEights commented on boards/raspberrypi/rpi_pico/rpi_pico_rp2040_w_defconfig:
```diff
@@ -9,3 +9,23 @@ CONFIG_BUILD_OUTPUT_HEX=y
+CONFIG_AIROC_WIFI_BUS_SDIO=n
+CONFIG_AIROC_WIFI_BUS_SPI=y
+CONFIG_AIROC_WIFI_BUS_SPI_DATA_IRQ_MUX=y
+CONFIG_CYW43439=y
```
Curious - yet the NVRAM image is included in the result ...
... and curiouser,  is also selected, even though I didn't put it in the configuration for the Pico W:
```
sboylan@system76-pc:/media/sboylan/LaCie/zephyr/bare-test/build/zephyr$ grep "CYW4343" .config
# CONFIG_CYW4343W is not set
# CONFIG_CYW43438 is not set
CONFIG_CYW43439=y
CONFIG_CYW43439_MURATA_1YN=y
```

npal-cy replied:
yes looks like if you do not select any module it takes default one.. anyway need to double check if for Raspberry Pi Pico W we need to use some specific FW/CLM/Nvram...

ThreeEights replied:
@npal-cy - I have no way to verify the firmware for the CYW43439 on the Pico W other than loading it and see what happens.  Do you know if the versions used in the Raspberry Pi Pico SDK are the same as what's supplied in the binary blobs?

ThreeEights replied:
@npal-cy - Were you able to figure out if the default blobs are correct for the Pico W?

npal-cy replied:
the current one is good for CYW43439 on the Pico W. The recommendation is to use latest FW, but it is out of this PR, will update once update WHD.

ThreeEights committed:
manifest:  Update hal_infineon to latest version

Adds custom HAL support for SPI

Signed-off-by: Steve Boylan <stephen.boylan@beechwoods.com>

jukkar commented on boards/raspberrypi/rpi_pico/rpi_pico_rp2040_w_defconfig:
```diff
@@ -9,3 +9,23 @@ CONFIG_BUILD_OUTPUT_HEX=y
+CONFIG_WIFI=y
+CONFIG_NETWORKING=y
+CONFIG_NET_L2_ETHERNET=y
+CONFIG_NET_IPV6=n
+CONFIG_NET_IPV4=y
+CONFIG_NET_DHCPV4=y
```
These options belong to application and not to board default, I suggest these are removed.

ThreeEights replied:
Thanks for pointing that out.  You are quite correct:  none of the CONFIG_NET* options belong in the defconfig.

jukkar commented on drivers/wifi/infineon/Kconfig.airoc:
```diff
@@ -37,6 +54,17 @@ config AIROC_WIFI_CUSTOM
+config AIROC_WIFI_BUS_SPI_DATA_IRQ_MUX
+	bool "Data and IRQ GPIO is shared"
+	help
+		Enable logic when the SPI data GPIO is shared
+		with the host wake interrupt, which requires
+		switching GPIO modes for that pin.
```
Typically the help text is aligned with two spaces, please change this.

ThreeEights replied:
Fixed.

npal-cy commented on drivers/wifi/infineon/Kconfig.airoc:
```diff
@@ -4,19 +4,37 @@
 
+config AIROC_WIFI_BUS_SDIO
+	bool
+	default y if SDHC
```
we should use 'if $(DT_COMPAT_INFINEON_AIROC_WIFI),sd)'. we can have a case when in system will select SPI and SDHC...for some not AIROC propose...

the same for AIROC_WIFI_BUS_SPI

ThreeEights replied:
@npal-cy - True.  The logic now looks like this:
```
config AIROC_WIFI_BUS_SDIO
	bool
	default y
	depends on $(dt_compat_on_bus,$(DT_COMPAT_INFINEON_AIROC_WIFI),sd)
	select SDHC
	select SDIO_STACK
	help
	  Enable SDIO bus support.

config AIROC_WIFI_BUS_SPI
	bool
	default y
	depends on $(dt_compat_on_bus,$(DT_COMPAT_INFINEON_AIROC_WIFI),spi)
	select SPI
	help
	  Enable SPI bus support
```

npal-cy commented on drivers/wifi/infineon/Kconfig.airoc:
```diff
@@ -4,19 +4,37 @@
-	depends on DT_HAS_INFINEON_AIROC_WIFI_ENABLED
+	select SDHC if $(dt_compat_on_bus,$(DT_COMPAT_INFINEON_AIROC_WIFI),sd)
+	select SDIO_STACK if $(dt_compat_on_bus,$(DT_COMPAT_INFINEON_AIROC_WIFI),sd)
+	select SPI if $(dt_compat_on_bus,$(DT_COMPAT_INFINEON_AIROC_WIFI),spi)
```
move `select SDHC`, `select SDIO_STACK`, `select SPI` to  `config  AIROC_WIFI_BUS_SPI` and `config AIROC_WIFI_BUS_SDIO`

ThreeEights replied:
Done.

ThreeEights commented:
@parthitce - Are you satisfied with the changes I made in response to your review?

soburi commented on drivers/wifi/infineon/airoc_wifi.h:
```diff
@@ -34,11 +62,25 @@ struct airoc_wifi_data {
 };
 
+union airoc_wifi_bus {
+#if defined(CONFIG_AIROC_WIFI_BUS_SDIO)
```
Use DT_ANY_INST_ON_BUS_STATUS_OKAY.
Although this is probably not a common case, it is best to assume that there are two devices connected via sdio and spi.

soburi commented on drivers/wifi/infineon/airoc_wifi.h:
```diff
@@ -34,11 +62,25 @@ struct airoc_wifi_data {
+union airoc_wifi_bus {
+#if defined(CONFIG_AIROC_WIFI_BUS_SDIO)
+	const struct device *bus_sdio;
+#elif defined(CONFIG_AIROC_WIFI_BUS_SPI)
```
I don't think this is exclusive.
Although this is probably not a common case, it is best to assume that there are two devices connected via sdio and spi.

ThreeEights replied:
On the one hand, the Infineon AIROC driver only supports a single device.  On the other hand ... if they ever change that, fixing this will mean the airoc_wifi_bus definition will not have to change.  Done.

ThreeEights commented on dts/bindings/wifi/infineon,airoc-wifi-spi.yaml:
```diff
@@ -0,0 +1,38 @@
+      - 8
+      - 16
+      - 32
+    default: 8
```
Updated with explanation.

ThreeEights commented on boards/raspberrypi/rpi_pico/rpi_pico_rp2040_w_defconfig:
```diff
@@ -9,3 +9,12 @@ CONFIG_BUILD_OUTPUT_HEX=y
+
+# Increase memory areas for WiFi stack
+CONFIG_MAIN_STACK_SIZE=4096
+CONFIG_HEAP_MEM_POOL_SIZE=16384
```
Thanks for the pointer to that capability.  Looking at some other boards, I moved all the memory config out to Kconfig.defconfig.

ThreeEights commented on boards/raspberrypi/rpi_pico/rpi_pico_rp2040_w_defconfig:
```diff
@@ -9,3 +9,12 @@ CONFIG_BUILD_OUTPUT_HEX=y
+CONFIG_SHELL_STACK_SIZE=2560
+
+# Default networking configuration
+CONFIG_AIROC_WIFI_COUNTRY="US"
```
Another vestige of long-gone changes.  Thanks for spotting that.

ajf58 commented on boards/raspberrypi/rpi_pico/rpi_pico_rp2040_w.dts:
```diff
@@ -1,9 +1,67 @@
+&pio0 {
+	status = "okay";
+
+	pio0_spi0: pio0_spi0 {
```
Very minor nit/suggestion (should not be a blocker): Could/should the device tree order of properties match the Linux DTS coding style? https://docs.kernel.org/devicetree/bindings/dts-coding-style.html#order-of-properties-in-device-node

ThreeEights replied:
That's a good suggestion.  If it becomes necessary to make any required changes for this pull request, I'll do that.

ajf58 commented on drivers/wifi/infineon/airoc_whd_hal_common.c:
```diff
@@ -0,0 +1,84 @@
+LOG_MODULE_DECLARE(infineon_airoc_wifi, CONFIG_WIFI_LOG_LEVEL);
+
+#ifdef __cplusplus
+extern "C" {
```
nit (not a blocker): Is this required?

ThreeEights replied:
Probably not.  In fact, a purist might say that it's wrong to have that in a C source file;  it only belongs in header files.  However, one of my goals was minimum disruption of the original sources ... though that seems to have fallen by the wayside.  The sources airoc_whd_hal_common.c, airoc_whd_hal_sdio.c, and airoc_whd_hal_spi.c were derived from the original airoc_whd_hal.c, and I didn't think about the presence of that #ifdef.  While I'm pretty sure that's superfluous, I'm reluctant to remove it at this late date.  Only if someone insists.

ajf58 replied:
This rationale makes sense to me., but I can't resolve the conversation (I would if I could).

ajf58 commented on drivers/wifi/infineon/airoc_whd_hal_common.c:
```diff
@@ -0,0 +1,84 @@
+#endif
+
+/******************************************************
+ *                 Function
```
nit: Is this required? It seems a bit vestigial.

ThreeEights replied:
Hmm ... true;  that also came from the original source file, and probably ought to be removed.

ajf58 commented on drivers/wifi/infineon/airoc_whd_hal_spi.c:
```diff
@@ -0,0 +1,289 @@
+
+	ret = spi_transceive_dt(spi_obj, &tx_set, &rx_set);
+	if (ret) {
+		LOG_DBG("spi_transceive FAIL %d\n", ret);
```
A question (not a blocker): Is this the correct log level, given the message's content?

ThreeEights replied:
Yeah, those should all be LOG_ERR.

ajf58 commented on drivers/wifi/infineon/airoc_wifi.h:
```diff
@@ -6,16 +6,44 @@
+#ifdef CONFIG_AIROC_WIFI_BUS_SDIO
+#include <zephyr/sd/sd.h>
+#include <zephyr/sd/sdio.h>
+#endif
+#ifdef CONFIG_AIROC_WIFI_BUS_SPI
```
```suggestion
#elif defined(CONFIG_AIROC_WIFI_BUS_SPI)
```
nit/suggestion: Would having the above suggestion making it clearer that the driver supports either SDIO, or SPI, but not both? 

This would match the implementation in drivers/wifi/infineon/airoc_wifi.c .

soburi replied:
If multiple instances cannot coexist due to essential(hardware) constraints, it is better, but because it's not, in this case, it has no benefit other than making it easier to understand, so I think it's fine to leave it as is.
If you want to fix it, it would be better to add support for multiple instances, but that doesn't need to be included in the scope of this PR.

ajf58 commented on drivers/wifi/infineon/airoc_wifi.h:
```diff
@@ -34,11 +62,26 @@ struct airoc_wifi_data {
+union airoc_wifi_bus {
+#if defined(CONFIG_AIROC_WIFI_BUS_SDIO)
+	const struct device *bus_sdio;
+#endif
+#if defined(CONFIG_AIROC_WIFI_BUS_SPI)
```
Same suggestion as above. Is this an either/or thing, or have I misunderstood the implementation?

ThreeEights replied:
Likewise.

ajf58 reviewed (approved):
I've left a couple of nits/suggestions.

I think the driver can only support SDIO or SPI (not both, concurrently). If that's correct, then maybe there's a a bit of scope for clarifying this in the #ifdef usage. Otherwise, I'm just being an idiot and misunderstanding.

Ditto I'm a bit surprised by the need for:
```
#ifdef __cplusplus
extern "C" {
#endif
```
in drivers/wifi/infineon/airoc_whd_hal_spi.c . I suspect they're innocuous in practice, but maybe there's some C++ code elsewhere that needs this.

Generally looks OK, and I assume it's been tested on target hardware so LGTM.

kartben commented:
@npal-cy @parthitce please re-visit

parthitce commented on drivers/wifi/infineon/airoc_whd_hal_spi.c:
```diff
@@ -0,0 +1,289 @@
+
+	/* Pull bus select line low before enabling WiFi chip */
+	gpio_pin_configure_dt(&config->bus_select_gpio, GPIO_OUTPUT_INACTIVE);
+	gpio_pin_set_dt(&config->bus_select_gpio, 0);
```
`GPIO_OUTPUT_INACTIVE` should put the pin in low state right?

ThreeEights replied:
Yep.  Good catch.

ThreeEights replied:
Fixed.

parthitce commented on drivers/wifi/infineon/airoc_whd_hal_spi.c:
```diff
@@ -0,0 +1,289 @@
+
+	if (airoc_wifi_power_on(dev)) {
+		LOG_ERR("airoc_wifi_power_on returns fail");
+		return -ENODEV;
```
IMO it makes sense to propagate the error number from `airoc_wifi_power_on`

ThreeEights replied:
I don't think so.  The errors from `airoc_wifi_power_on` are low-level GPIO errors, which are meaningless to applications trying to use WiFi.  I'd say it makes sense here to wrap those in the (slightly) more meaningful ENODEV.

parthitce commented on drivers/wifi/infineon/airoc_whd_hal_spi.c:
```diff
@@ -0,0 +1,289 @@
+		.is_spi_normal_mode = WHD_FALSE,
+	};
+
+#if DT_INST_NODE_HAS_PROP(0, wifi_host_wake_gpios)
```
this section of code is both in spi and sdio now.

parthitce commented on drivers/wifi/infineon/airoc_whd_hal_spi.c:
```diff
@@ -0,0 +1,289 @@
+	whd_result_t whd_ret = WHD_SUCCESS;
+
+#if defined(SPI_DATA_IRQ_SHARED)
+	ret = whd_bus_spi_irq_enable(whd_driver, false);
```
SPI operations needs to be done with interrupt disabled?

ThreeEights replied:
Yes, if the same SPI line is being used for data transfers and interrupts, which is the case on the Raspberry Pi Pico W.  Interrupts need to be disabled so data transfers don't trigger the interrupt.

parthitce replied:
Driver is specific to board / SoC then? It shouldn't AFAIK.

ThreeEights replied:
No, though it currently applies to just two boards, the Raspberry Pi Pico W and Pico 2 W.  Given Zephyr's code layers, the SPI driver has no way to know the WiFi module has this wacky wiring configuration - nor should it.  It's specific to WiFi on these boards, so it's the WiFi driver layer that has to deal with the issue, but anything in the WiFi driver to do so should otherwise be board-agnostic.  This was the best thing I could think of to encapsulate the problem.

parthitce replied:
If this modules is going to be used in other boards, then it makes sense to add an additional quirk from dts (something like shared-interrupt) and differentiate it. Otherwise this module is bound to RPi pico families.

ThreeEights replied:
I believe you're suggesting something like the property `spi-data-irq-shared`, which is defined in dts/bindings/wifi/infineon,airoc-wifi-spi.yaml?

parthitce replied:
Oh, I missed that. yes can `spi-data-irq-shared` help make this part of code independent of Pico?

ThreeEights replied:
That was my goal:  any device that needs to use a single pin for data input and interrupts from the WiFi module can use that property.  You can see it used in the board definition for the Pico W:  boards/raspberrypi/rpi_pico/rpi_pico_rp2040_w.dts.

Originally I had Pico-specific code here, and previous reviewers pointed out how inappropriate that was, leading me to figure out a platform-independent way to trigger the necessary logic.

parthitce replied:
But still the underlying idea about disabling interrupt in SPI as it's shared with the data pin is here?

ThreeEights replied:
I don't understand your question, so let me restate the problem and my solution:  If the same GPIO pin is used for both data and interrupts coming from the WiFi module, interrupts have to be disabled during data transfers to prevent the data on the SPI input from triggering interrupts.  That's not a function of the SPI bus layer, it belongs in the layer managing WiFi transfers;  here, that's the Infineon AIROC driver.  Though there are currently only two boards that have this issue, the Raspberry Pi Pico W and Pico 2 W, I tried to implement this in a way that wasn't specific to those boards.

Can you explain what your concern is about this solution?

parthitce replied:
You have the conditional ennoblement of interrupt using `#if defined(SPI_DATA_IRQ_SHARED)`, but do we still need to disable the interrupt even if data and interrupt is not shared -> https://github.com/zephyrproject-rtos/zephyr/pull/78773/files#diff-a359abf10a98428aea5f07434fb65f88e4544a4ddf9d26029ee3739c65fedea6R271 ?

ThreeEights replied:
I think I understand your question now.  Enabling and disabling the IRQ line is normally under control of the driver code in the Infineon HAL module (modules/hal/infineon).  The function whd_bus_spi_irq_enable() (and the corresponding whd_bus_sdio_irq_enable() and whd_bus_m2m_irq_enable()) in the drivers/wifi/infineon layer is the platform-specific bit to do so.  That logic is independent of whether or not the GPIO is shared.

I think the board designers at Raspberry Pi were getting a bit desperate to squeeze the CYW43439 onto their board without disrupting the rest of their I/O design.  I hope no other board designer has to deal with a situation like this.

ajf58 commented on boards/raspberrypi/rpi_pico/doc/index.rst:
```diff
@@ -153,6 +153,23 @@ devices as well as both PIO devices).
+Prerequisites for the Pico W
+----------------------------
+
+Building for the RaspberryPi Pico W requires the AIROC binary blobs
```
Not  a blocker:
```suggestion
Building for the Raspberry Pi Pico W requires the AIROC binary blobs
```

ThreeEights replied:
Something needed to change, so this is going in.

ajf58 reviewed (approved):
LGTM.

All my comments/suggestions have been answered and/or corrected. I spotted a couple of tiny tweaks (sorry I didn't spot them earlier), but definitely should not blocker such a significant improvement to in-tree functionality for this board.

jukkar commented:
@parthitce @npal-cy please re-review

ThreeEights commented on drivers/wifi/infineon/CMakeLists.txt:
```diff
@@ -5,7 +5,9 @@
 
 zephyr_library_sources_ifdef(CONFIG_WIFI_AIROC airoc_wifi.c)
-zephyr_library_sources_ifdef(CONFIG_WIFI_AIROC airoc_whd_hal.c)
+zephyr_library_sources_ifdef(CONFIG_WIFI_AIROC airoc_whd_hal_common.c)
```
Good point.  I'll add that next time something needs to change.

ThreeEights replied:
Something needed to change, so this is going in.

npal-cy commented on drivers/wifi/infineon/Kconfig.airoc:
```diff
@@ -4,20 +4,40 @@
+	help
+	  Enable SPI bus support
+
+config AIROC_WIFI_COUNTRY
```
where do we use AIROC_WIFI_COUNTRY?

I see we also has CY_WIFI_COUNTRY, looks like we can AIROC_WIFI_COUNTRY instead CY_WIFI_COUNTRY (and remove CY_WIFI_COUNTRY)

parthitce reviewed (approved):
CI failures needs to be addressed. Otherwise LGTM

ifyall commented:
@ThreeEights, I am just waiting on some functional tests from my team on the existing Wi-Fi support with this change. Once that is done and is passing, I will approve.

Ian

ThreeEights commented:
@ifyall - Thanks!

jukkar commented:
ping @npal-cy please take a look

npal-cy commented:
> ping @npal-cy please take a look
I am going to do some tests today, before merge...

kartben commented:
Yay!!! Thanks for the hard work @ThreeEights!

ThreeEights commented:
@ifyall @npal-cy @yonsch @parthitce @kartben @jukkar @soburi @ajf58 - Thanks for all your help making this clean and functional.
